### PR TITLE
RIT-289

### PR DESCRIPTION
**Pull request by jrocca82**

### CODE REVIEW

File: contracts/RitestreamNFT.sol [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import "@openzeppelin/contracts/utils/Strings.sol";
+
+contract RitestreamNFT is ERC721Enumerable, Ownable {
+    using Strings for uint256;
+    using Counters for Counters.Counter;
+
+    Counters.Counter private blueTokenCounter;
+    Counters.Counter private greenTokenCounter;
+    Counters.Counter private redTokenCounter;
+
```
Comment (xiangzhe0812) [inline contracts/RitestreamNFT.sol:16]: can be removed

File: contracts/RitestreamNFT.sol [hunk 2]
```diff
@@ -0,0 +1,113 @@
+pragma solidity ^0.8.3;
+pragma abicoder v2;
+
+import "@openzeppelin/contracts/token/ERC721/extensions/ERC721Enumerable.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/Counters.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
```
Comment (xiangzhe0812) [inline contracts/RitestreamNFT.sol:7]: can be removed

File: contracts/RitestreamNFT.sol [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+    function nextGreenTokenId() internal view returns (uint256) {
+        uint256 greenTokenId = greenTokenCount + 1;
+        return greenTokenId;
+    }
+
+    //FUNCTION FOR MINTING
+    function mintBlueTokens(address userAddress) external saleActive onlyOwner {
+        require(
+            blueTokenCount <= blueMax,
```
Comment (rellfy) [inline contracts/RitestreamNFT.sol:82]:
```suggestion
            blueTokenCount < blueMax,
```

File: contracts/RitestreamNFT.sol [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        blueTokenCount += 1;
+    }
+
+    function mintGreenTokens(address userAddress)
+        external
+        saleActive
+        onlyOwner
+    {
+        require(
+            greenTokenCount <= greenMax,
```
Comment (rellfy) [inline contracts/RitestreamNFT.sol:95]:
```suggestion
            greenTokenCount < greenMax,
```

File: contracts/RitestreamNFT.sol [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            "Not enough green passes remaining to mint"
+        );
+        _safeMint(userAddress, nextGreenTokenId());
+        greenTokenCount += 1;
+    }
+
+    function mintRedTokens(address userAddress) external saleActive onlyOwner {
+        require(
+            redTokenCount <= redMax,
```
Comment (rellfy) [inline contracts/RitestreamNFT.sol:104]:
```suggestion
            redTokenCount < redMax,
```

**Review by rellfy — DISMISSED** [review]
Need to change the LTE to LT so minting doesnt exceed the max

File: tests/RitestreamNFT.spec.ts [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    }
+  });
+
+  it("Should mint blue tokens", async () => {
+    await token.mintBlueTokens(userAddress);
+    const balance = await token.balanceOf(userAddress);
+    expect(balance).to.equal(2);
+
+    //Check token ID has increased:
+    expect(token.nextBlueTokenId === 2);
```
Comment (bitcoinbrisbane) [inline tests/RitestreamNFT.spec.ts:77]: I think its better to do
```suggestion
    expect(token.nextBlueTokenId).to.eq(2);
```

File: contracts/RitestreamNFT.sol [hunk 7]
```diff
@@ -0,0 +1,119 @@
+pragma solidity ^0.8.3;
+pragma abicoder v2;
+
+import "@openzeppelin/contracts/token/ERC721/extensions/ERC721Enumerable.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+contract RitestreamNFT is ERC721Enumerable, Ownable {
+    address public immutable _self;
```
Comment (bitcoinbrisbane) [inline contracts/RitestreamNFT.sol:8]: I think this should be private.  It will save gas if it is.
```suggestion
    address private immutable _self;
```

**Review by bitcoinbrisbane — CHANGES_REQUESTED** [review]
Fair few things to look at

File: contracts/RitestreamNFT.sol [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        return redTokenId;
+    }
+
+    function nextGreenTokenId() internal view returns (uint256) {
+        uint256 greenTokenId = greenTokenCount + 1;
+        return greenTokenId;
+    }
+
+    //FUNCTION FOR MINTING
+    function mintBlueTokens(address userAddress) external saleActive onlyOwner {
```
Comment (xiangzhe0812) [inline contracts/RitestreamNFT.sol:84]:
```suggestion
    function mintBluePasses(address userAddress) external saleActive onlyOwner {
```

File: contracts/RitestreamNFT.sol [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        require(
+            blueTokenCount < blueMax,
+            "Not enough blue passes remaining to mint"
+        );
+        _safeMint(userAddress, nextBlueTokenId());
+        blueTokenCount += 1;
+    }
+
+    function mintGreenTokens(address userAddress)
```
Comment (xiangzhe0812) [inline contracts/RitestreamNFT.sol:93]:
```suggestion
    function mintGreenPasses(address userAddress)
```

File: contracts/RitestreamNFT.sol [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    {
+        require(
+            greenTokenCount < greenMax,
+            "Not enough green passes remaining to mint"
+        );
+        _safeMint(userAddress, nextGreenTokenId());
+        greenTokenCount += 1;
+    }
+
+    function mintRedTokens(address userAddress) external saleActive onlyOwner {
```
Comment (xiangzhe0812) [inline contracts/RitestreamNFT.sol:106]:
```suggestion
    function mintRedPasses(address userAddress) external saleActive onlyOwner {
```

File: tests/RitestreamNFT.spec.ts [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+  });
+
+  it("Should mint red tokens", async () => {
```
Comment (xiangzhe0812) [inline tests/RitestreamNFT.spec.ts:80]:
```suggestion
  it("Should mint red passes", async () => {
```

File: tests/RitestreamNFT.spec.ts [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    await token.mintRedTokens(userAddress);
+    const balance = await token.balanceOf(userAddress);
+    expect(balance).to.equal(3);
+
+    //Check token ID has increased:
+    expect(token.nextRedTokenId === 4002);
+  });
+
+  it("Should mint green tokens", async () => {
```
Comment (xiangzhe0812) [inline tests/RitestreamNFT.spec.ts:89]:
```suggestion
  it("Should mint green passes", async () => {
```

File: tests/RitestreamNFT.spec.ts [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    await token.mintGreenTokens(userAddress);
+    const balance = await token.balanceOf(userAddress);
+    expect(balance).to.equal(4);
+
+    //Check token ID has increased:
+    expect(token.nextGreenTokenId === 7002);
+  });
+
+  //Time out in for loop if trying to test other colors, but logic is same
+  it("Should not mint more than 1000 green tokens", async () => {
```
Comment (xiangzhe0812) [inline tests/RitestreamNFT.spec.ts:99]:
```suggestion
  it("Should not mint more than 1000 green passes", async () => {
```

**Review by xiangzhe0812 — COMMENTED** [review]
few more comments.

**Review by xiangzhe0812 — DISMISSED** [review]

File: contracts/RitestreamNFT.sol (hunk 10 above)
Comment (bitcoinbrisbane) [reply contracts/RitestreamNFT.sol]: I think this would be more scalable if we used an enum. And the contract would be smaller.

**Comment by bitcoinbrisbane** [comment]
A fair few things need attention.

- [ ] Rename dev.env to our standards
- [ ] Inconsistent solidity version
- [ ] No //SPDX licence on some contracts, needs to be MIT
- [ ] No MIT licence in package.json
- [ ] Not sure why we are using our own IERC20.sol.  Use OpenZepelin  https://github.com/dltxio/ritestream-contract/blob/master/contracts/interfaces/IERC20.sol like here https://github.com/dltxio/ritestream-contract/blob/master/contracts/Vault.sol#L4
- [ ] Not sure what IWETH.sol is
- [ ] Check this guide for layouts, needs to be agreed in house

**Review by bitcoinbrisbane — CHANGES_REQUESTED** [review]
See comments

**Review by bitcoinbrisbane — DISMISSED** [review]

File: contracts/TeamVesting.sol [hunk 14]
```diff
@@ -1,4 +1,5 @@
-pragma solidity 0.8.4;
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
```
Comment (xiangzhe0812) [inline contracts/TeamVesting.sol:2]:
```suggestion
pragma solidity 0.8.4;
```
using the fixed solidity version was required by audit team, lets use 0.8.4 and keep all the same

File: contracts/Token.sol [hunk 15]
```diff
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: MIT
-pragma solidity 0.8.4;
+pragma solidity ^0.8.4;
```
Comment (xiangzhe0812) [inline contracts/Token.sol:2]:
```suggestion
pragma solidity 0.8.4;
```
using the fixed solidity version was required by audit team, lets use 0.8.4 and keep all the same

File: contracts/Vault.sol [hunk 16]
```diff
@@ -1,4 +1,5 @@
-pragma solidity 0.8.4;
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
```
Comment (xiangzhe0812) [inline contracts/Vault.sol:2]:
```suggestion
pragma solidity 0.8.4;
```
using the fixed solidity version was required by audit team, lets use 0.8.4 and keep all the same

File: contracts/RitestreamNFT.sol [hunk 17]
```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
```
Comment (xiangzhe0812) [inline contracts/RitestreamNFT.sol:2]:
```suggestion
pragma solidity 0.8.4;
```
using the fixed solidity version was required by audit team, lets use 0.8.4 and keep all the same

**Review by xiangzhe0812 — DISMISSED** [review]
need to use the fixed solidity version

**Review by xiangzhe0812 — APPROVED** [review]

**Review by ahinchliff — APPROVED** [review]

**Pull request merged**